### ISSUE TITLE
foreach_dataset should be able to operate on super only

### ISSUE THREAD

**Issue by bpoldrack** [datalad/datalad#6659]
While it's not instantly obvious from a looking at it as a stand-alone command, `foreach-dataset` should be able to run things on the uppermost dataset only. That boils down to programmatic uses that include passing on `recursive`. Right now, `foreach-dataset` (and ultimately `subdatasets`) are looking into/running on the first level subdatasets with `recursive=False`.

Example usecase is in `create-sibling-webdav` in datalad-next.

A solution could be to switch the defaults:

Now: `recursive=False`, `recursion_limit=None`.

Could possibly be: `recursive=True`, `recursion_limit=1`. 

That should result in no change to the current default behavior when using it standalone, but allows callers to not treat their own `recursive` like that:

```
if not recursive:
    do_something(ds)
else:
    ds.foreach_dataset(do_somthing, recursive=recursive, recursion_limit=recursion_limit
```

and simply have that `else` block only.

**Comment by mih** [comment]
As shown in https://github.com/datalad/datalad-next/pull/21/files#r859565784, non-recursive behavior needs `datalad foreach-dataset -R 0` right now

**Comment by yarikoptic** [comment]
Just to clear up: you would like to use `foreach-dataset` directly on a singular dataset to benefit from all the goodies of passing various `{format}` variables into your call? or what is the reason for such a desire for a command need for which was initially to operate across multiple datasets.

**Comment by bpoldrack** [comment]
@yarikoptic 
> Just to clear up: you would like to use foreach-dataset directly on a singular dataset to benefit from all the goodies of passing various {format} variables into your call? or what is the reason for such a desire for a command need for which was initially to operate across multiple datasets.

Yes and no. The point is, that my code calling `foreach-dataset` is parameterized by `recursive` itself. So I don't know yet whether my function needs to be executed on top-ds only or recursively.

**Comment by yarikoptic** [comment]
ok, I guess I got a bit mislead by the issue title "foreach_dataset should be able to operate on super only" since the answer to it is really give by @mih -- just use `-R 0` option.

The actual issue (OP description) is about a somewhat inconsistent with the other tools notion of `recursive`, that with `recursive=False` (default) `foreach-dataset` operates not only on the super, but also on immediate sub-datasets.  Thus unlike `datalad subdatasets`  (doesn't include super) it pretty much includes two levels in the hierarchy (super + immediate subdatasets). (note: there is an extra option `-s, --subdatasets-only` to work on only subdatasets, no super).

I think that the suggested by @bpoldrack change of defaults to `recursive=True, recursion_limit=1` could be a cute way out which would retain prior default behavior, while more natively providing a way to be used in other functionality.  So far I do not see possible side-effect and would welcome a PR. The only minor "concern" is that such defaults would be somewhat "non-conventional".

Another approach could be to change behavior of `foreach-dataset` to not cross boundaries of the hierarchy by default and to not operate on immediate subdatasets. Then I think it could also be natively embedded to pass those recursion options. To replicate previous behavior it would need that `-r -R1` .  Going wilder, we might want to generalize `-s, --subdatasets-only` to `-L, --recursion-low-limit` so `-r -L1 -R1` would limit to only subdatasets at level 1, and `-r -L2 -R4` would skip immediate subdatasets (e.g. top level datasets, and operate on two next levels).  But all that might be not yet warranted and not worth breaking the released behavior. WDYT?

**Comment by bpoldrack** [comment]
While obviously I kinda like my suggestion, we should be aware that it still breaks behavior. If a caller explicitly specified `recursive=True` but no recursion limit before, this would behave differently and somewhat unexpectedly (as you said - "non-conventional"). However, `recursion_limit` being a meaningful option when `recursive` was `False` feels "non-conventional", too - not sure that works properly across all commands.

Re `recursion-low-limit`:

Wild indeed. I lean towards agreement on this:

>  But all that might be not yet warranted and not worth breaking the released behavior.

**Comment by yarikoptic** [comment]
I remember writing a reply but I guess I failed to press "Comment" ;)  1. Thanks @bpoldrack  -- I missed that; 2. I am leaning toward unifying `foreach-dataset` behavior as I have described [above](https://github.com/datalad/datalad/issues/6659#issuecomment-1112537842) even though indeed it would be a breaking change: I should have thought better to not introduce "cross-levels" behavior by default in that command, and stayed consistent with the rest.
